### PR TITLE
Refactor pipeline parallel device config

### PR DESCRIPTION
**Pull request by sgwhat**
## Description

This PR is mainly created to optimize the usage for pipeline parallel inference.

### 1. Why the change?

There are too many configurations (such as device_map and model_layers) that need to be specified by the user in the example.

### 2. User API changes
```python
    model = AutoModelForCausalLM.from_pretrained(model_path,
                                                 load_in_4bit=True,
                                                 optimize_model=True,
                                                 trust_remote_code=True,
                                                 use_cache=True,
                                                 pipeline_parallel_stages=2)
```

### 4. How to test?
- [ ] Unit test
- [x] Application test

### CODE REVIEW

File: python/llm/example/GPU/Pipeline-Parallel-Inference/generate.py [hunk 1]
```diff
@@ -62,27 +62,9 @@ def get_prompt(message: str, chat_history: list[tuple[str, str]],
...
-    from accelerate import dispatch_model
-    model = dispatch_model(
-        model,
-        device_map=device_map,
-        offload_dir=None,
-        skip_keys=["past_key_value", "past_key_values"],
-    )
+                                                 use_cache=True,
+                                                 pipeline_parallel=True,
+                                                 parallel_gpu_num=args.gpu_num)
```
Comment (jason-dai) [inline python/llm/example/GPU/Pipeline-Parallel-Inference/generate.py:67]: remove `pipeline_parallel` and change to `pipeline_parallel_stages=args.gpu_num`

File: python/llm/src/ipex_llm/transformers/model.py [hunk 2]
```diff
@@ -157,6 +157,10 @@ def from_pretrained(cls,
         :param mixed_precision: boolean value, Whether to use mixed precision quantization.
             Default to be False. If set to True, we will use sym_int8 for lm_head when
             load_in_low_bit is sym_int4 or asym_int4.
+        :param pipeline_parallel: boolean value, Whether to use pipeline parallel inference.
+                                  Default to be ``False``.
+        :param parallel_gpu_num: int value, the number of GPUs allocated for pipeline parallel
+                                 inference. Default to be ``None``.
```
Comment (plusbang) [inline python/llm/src/ipex_llm/transformers/model.py:163]: I feel it maybe better that default to be 1.

File: python/llm/src/ipex_llm/transformers/model.py [hunk 3]
```diff
@@ -346,6 +352,26 @@ def from_pretrained(cls,
...
+                    device_map.update({key: f'xpu:{i}' for key in
+                                       model_layers[split_len * i: split_len * (i + 1)]})
+                    if i == parallel_gpu_num - 1:
+                        device_map.update({key: f'xpu:{i}' for key in
+                                           model_layers[split_len * (i + 1):]})
+
+                from accelerate import dispatch_model
+                model = dispatch_model(
+                    model, device_map=device_map, skip_keys=["past_key_value", "past_key_values"],
+                )
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/model.py:373]: Maybe put into another function `pipeline_parallelize(model)`?

File: python/llm/src/ipex_llm/transformers/model.py [hunk 4]
```diff
@@ -346,6 +352,26 @@ def from_pretrained(cls,
...
+
             if speculative:
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/model.py:375]: Is `speculative` compatible with pipeline parallel?

**Review by plusbang — COMMENTED** [review]
Please also update usage in all-in-one benchmark script: https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/dev/benchmark/all-in-one/run.py#L1688

File: python/llm/src/ipex_llm/transformers/model.py (hunk 4 above)
Comment (sgwhat) [reply python/llm/src/ipex_llm/transformers/model.py:375]: no, it's not compatible now.

Comment (jason-dai) [reply python/llm/src/ipex_llm/transformers/model.py:375]: Then we need to check and report error when both are enabled.

Comment (sgwhat) [reply python/llm/src/ipex_llm/transformers/model.py:375]: Do we need to add the support for enabling both speculative and pipeline parallel later?

Comment (jason-dai) [reply python/llm/src/ipex_llm/transformers/model.py:375]: Maybe not critical

File: python/llm/src/ipex_llm/transformers/model.py (hunk 2 above)
Comment (sgwhat) [reply python/llm/src/ipex_llm/transformers/model.py]: this script will be updated in a later pr.

File: python/llm/src/ipex_llm/transformers/model.py [hunk 5]
```diff
@@ -190,6 +215,7 @@ def from_pretrained(cls,
         optimize_model = kwargs.pop("optimize_model", True)
         user_quantization_config = kwargs.pop("quantization_config", None)
         speculative = kwargs.pop("speculative", False)
+        pipeline_parallel_stages = kwargs.pop("pipeline_parallel_stages", None)
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/model.py:218]: default to 1

Comment (sgwhat) [reply python/llm/src/ipex_llm/transformers/model.py]: done.

**Review by jason-dai — APPROVED** [review]
LGTM

**Pull request merged**